File: src/cs/PhantasyIslandPythonRemoteControl/radio/Apis.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;

namespace PhantasyIslandPythonRemoteControl.Radio
{
    public class DebugApi : ApiModule
    {
        public DebugApi(RadioManager rm) : base(rm)
        {
        }

        public object Ping() => SendInternal("ping", waitCmd: "pong");
    }

    public class SceneApi : ApiModule
    {
        public SceneApi(RadioManager rm) : base(rm)
        {
        }

        public object ListAllMeshObjectInScene() => SendInternal("scene.listAllMeshObjectInScene");

        public object GetObjectInfoById(string objectId) => SendInternal("scene.getObjectInfoById",
            new Dictionary<string, object> { ["objectId"] = objectId });

        public object RemoveObjectById(string objectId) => SendInternal("scene.removeObjectById",
            new Dictionary<string, object> { ["objectId"] = objectId });

        public object MoveObjectById(string objectId, XYZ position) => SendInternal("scene.moveObjectById",
            new Dictionary<string, object> { ["objectId"] = objectId, ["position"] = position.ToList() });

        public object SetObjectRadioMaterial(string objectId, string materialId, double? thickness_m) => SendInternal(
            "scene.setObjectRadioMaterial",
            new Dictionary<string, object>
                { ["objectId"] = objectId, ["materialId"] = materialId, ["thickness_m"] = thickness_m });

        public object UpdateMeshViewMaterial(string meshId, ViewMaterialChangeCommand command) => SendInternal(
            "scene.updateMeshViewMaterial",
            new Dictionary<string, object> { ["meshId"] = meshId, ["viewMaterialChangeCommand"] = command.ToDict() });

        public object UpdateMeshViewMaterialSimple(string meshId, ViewMaterialChangeCommandSimple command) =>
            SendInternal("scene.updateMeshViewMaterialSimple",
                new Dictionary<string, object>
                    { ["meshId"] = meshId, ["viewMaterialChangeCommandSimple"] = command.ToDict() });
    }

    public class FlyApi : ApiModule
    {
        public FlyApi(RadioManager rm) : base(rm)
        {
        }

        public object ListFlyObject() => SendInternal("fly.listFlyObject");

        public object GetFlyObjectInfo(string keyName) => SendInternal("fly.getFlyObjectInfo",
            new Dictionary<string, object> { ["keyName"] = keyName });

        public object GetFlyObjectCameraImageDown(string keyName) => SendInternal("fly.getFlyObjectCameraImageDown",
            new Dictionary<string, object> { ["keyName"] = keyName });

        public object GetFlyObjectCameraImageFront(string keyName) => SendInternal("fly.getFlyObjectCameraImageFront",
            new Dictionary<string, object> { ["keyName"] = keyName });
    }

    public class RadioApi : ApiModule
    {
        public RadioApi(RadioManager rm) : base(rm)
        {
        }

        public object IsSceneInit() =>
            SendInternal("radio.isSceneInit", postProcessor: d => d.GetProperty("init").GetBoolean());

        public object IsRadioReachabilityCheckerInit() => SendInternal("radio.isRadioReachabilityCheckerInit",
            postProcessor: d => d.GetProperty("init").GetBoolean());

        public object CheckReachability(XYZ aTx, XYZ bRx, RadioCheckOptions options) => SendInternal(
            "radio.checkReachability",
            new Dictionary<string, object>
                { ["aTx"] = aTx.ToList(), ["bRx"] = bRx.ToList(), ["options"] = options?.ToDict() });

        public object UpdateObjectPos(string objectId, XYZ position) => SendInternal("radio.updateObjectPos",
            new Dictionary<string, object> { ["objectId"] = objectId, ["position"] = position.ToList() });

        public object GetObjectPos(string objectId) => SendInternal("radio.getObjectPos",
            new Dictionary<string, object> { ["objectId"] = objectId },
            postProcessor: d => d.GetProperty("position").EnumerateArray().Select(n => n.GetDouble()).ToList());

        public object UpdateMeshRadioMaterial(string meshId, string materialId, double? thickness_m) => SendInternal(
            "radio.updateMeshRadioMaterial",
            new Dictionary<string, object>
                { ["meshId"] = meshId, ["materialId"] = materialId, ["thickness_m"] = thickness_m });

        public object GetAllRadioMaterial() => SendInternal("radio.getAllRadioMaterial",
            postProcessor: d => d.GetProperty("meshIds").EnumerateArray().Select(n => ParseMaterial(n)).ToList());

        public object LocalRadioMaterial() => SendInternal("radio.localRadioMaterial",
            postProcessor: d => d.GetProperty("meshIds").EnumerateArray().Select(n => ParseMaterial(n)).ToList());

        public object GetBuildingRadioMaterial() => SendInternal("radio.getBuildingRadioMaterial");

        public object AddRadioMaterial(RadioMaterialProperties material) =>
            SendInternal("radio.addRadioMaterial", material.ToDict());

        public object ListRadioLocalObjectsIds() => SendInternal("radio.listRadioLocalObjects",
            postProcessor: d =>
                d.TryGetProperty("localObjectIds", out var ids)
                    ? ids.EnumerateArray().Select(n => n.GetString()).ToList()
                    : new List<string>());

        private RadioMaterialProperties ParseMaterial(JsonElement n) => new RadioMaterialProperties
        {
            Id = n.GetProperty("id").GetString(),
            DisplayName = n.GetProperty("displayName").GetString(),
            PenetrationLoss_dBPerMeter = n.GetProperty("penetrationLoss_dBPerMeter").GetDouble(),
            ReflectionCoefficient = n.GetProperty("reflectionCoefficient").GetDouble(),
            DefaultThickness_m = n.GetProperty("defaultThickness_m").GetDouble()
        };
    }
}

File: src/fs/Apis.fs
namespace PhantasyIslandRemoteControl.Radio

open System
open System.Collections.Generic
open System.Text.Json
open System.Linq

type DebugApi(rm: IRadioManager) =
    inherit ApiModule(rm)
    member this.Ping() = this.SendInternal("ping", waitCmd = "pong")

type SceneApi(rm: IRadioManager) =
    inherit ApiModule(rm)
    
    member this.ListAllMeshObjectInScene() = this.SendInternal("scene.listAllMeshObjectInScene")

    member this.GetObjectInfoById(objectId: string) =
        let data = Dictionary<string, obj>()
        data.["objectId"] <- objectId
        this.SendInternal("scene.getObjectInfoById", data)

    member this.RemoveObjectById(objectId: string) =
        let data = Dictionary<string, obj>()
        data.["objectId"] <- objectId
        this.SendInternal("scene.removeObjectById", data)

    member this.MoveObjectById(objectId: string, position: XYZ) =
        let data = Dictionary<string, obj>()
        data.["objectId"] <- objectId
        data.["position"] <- position.ToList()
        this.SendInternal("scene.moveObjectById", data)

    member this.SetObjectRadioMaterial(objectId: string, materialId: string, thickness_m: Nullable<double>) =
        let data = Dictionary<string, obj>()
        data.["objectId"] <- objectId
        data.["materialId"] <- materialId
        data.["thickness_m"] <- if thickness_m.HasValue then thickness_m.Value :> obj else null
        this.SendInternal("scene.setObjectRadioMaterial", data)

    member this.UpdateMeshViewMaterial(meshId: string, command: ViewMaterialChangeCommand) =
        let data = Dictionary<string, obj>()
        data.["meshId"] <- meshId
        data.["viewMaterialChangeCommand"] <- command.ToDict()
        this.SendInternal("scene.updateMeshViewMaterial", data)

    member this.UpdateMeshViewMaterialSimple(meshId: string, command: ViewMaterialChangeCommandSimple) =
        let data = Dictionary<string, obj>()
        data.["meshId"] <- meshId
        data.["viewMaterialChangeCommandSimple"] <- command.ToDict()
        this.SendInternal("scene.updateMeshViewMaterialSimple", data)

type FlyApi(rm: IRadioManager) =
    inherit ApiModule(rm)
    
    member this.ListFlyObject() = this.SendInternal("fly.listFlyObject")

    member this.GetFlyObjectInfo(keyName: string) =
        let data = Dictionary<string, obj>()
        data.["keyName"] <- keyName
        this.SendInternal("fly.getFlyObjectInfo", data)

    member this.GetFlyObjectCameraImageDown(keyName: string) =
        let data = Dictionary<string, obj>()
        data.["keyName"] <- keyName
        this.SendInternal("fly.getFlyObjectCameraImageDown", data)

    member this.GetFlyObjectCameraImageFront(keyName: string) =
        let data = Dictionary<string, obj>()
        data.["keyName"] <- keyName
        this.SendInternal("fly.getFlyObjectCameraImageFront", data)

type RadioApi(rm: IRadioManager) =
    inherit ApiModule(rm)

    let parseMaterial (n: JsonElement) =
        let m = RadioMaterialProperties()
        m.Id <- n.GetProperty("id").GetString()
        m.DisplayName <- n.GetProperty("displayName").GetString()
        m.PenetrationLoss_dBPerMeter <- n.GetProperty("penetrationLoss_dBPerMeter").GetDouble()
        m.ReflectionCoefficient <- n.GetProperty("reflectionCoefficient").GetDouble()
        m.DefaultThickness_m <- n.GetProperty("defaultThickness_m").GetDouble()
        m

    member this.IsSceneInit() =
        this.SendInternal("radio.isSceneInit", postProcessor = Func<JsonElement, obj>(fun d -> d.GetProperty("init").GetBoolean() :> obj))

    member this.IsRadioReachabilityCheckerInit() =
        this.SendInternal("radio.isRadioReachabilityCheckerInit", postProcessor = Func<JsonElement, obj>(fun d -> d.GetProperty("init").GetBoolean() :> obj))

    member this.CheckReachability(aTx: XYZ, bRx: XYZ, options: RadioCheckOptions) =
        let data = Dictionary<string, obj>()
        data.["aTx"] <- aTx.ToList()
        data.["bRx"] <- bRx.ToList()
        data.["options"] <- if not (Object.ReferenceEquals(options, null)) then options.ToDict() :> obj else null
        this.SendInternal("radio.checkReachability", data)

    member this.UpdateObjectPos(objectId: string, position: XYZ) =
        let data = Dictionary<string, obj>()
        data.["objectId"] <- objectId
        data.["position"] <- position.ToList()
        this.SendInternal("radio.updateObjectPos", data)

    member this.GetObjectPos(objectId: string) =
        let data = Dictionary<string, obj>()
        data.["objectId"] <- objectId
        this.SendInternal("radio.getObjectPos", data, postProcessor = Func<JsonElement, obj>(fun d -> 
            d.GetProperty("position").EnumerateArray().Select(fun n -> n.GetDouble()).ToList() :> obj))

    member this.UpdateMeshRadioMaterial(meshId: string, materialId: string, thickness_m: Nullable<double>) =
        let data = Dictionary<string, obj>()
        data.["meshId"] <- meshId
        data.["materialId"] <- materialId
        data.["thickness_m"] <- if thickness_m.HasValue then thickness_m.Value :> obj else null
        this.SendInternal("radio.updateMeshRadioMaterial", data)

    member this.GetAllRadioMaterial() =
        this.SendInternal("radio.getAllRadioMaterial", postProcessor = Func<JsonElement, obj>(fun d ->
            d.GetProperty("meshIds").EnumerateArray().Select(fun n -> parseMaterial n).ToList() :> obj))

    member this.LocalRadioMaterial() =
        this.SendInternal("radio.localRadioMaterial", postProcessor = Func<JsonElement, obj>(fun d ->
            d.GetProperty("meshIds").EnumerateArray().Select(fun n -> parseMaterial n).ToList() :> obj))

    member this.GetBuildingRadioMaterial() = this.SendInternal("radio.getBuildingRadioMaterial")

    member this.AddRadioMaterial(material: RadioMaterialProperties) =
        this.SendInternal("radio.addRadioMaterial", material.ToDict())

    member this.ListRadioLocalObjectsIds() =
        this.SendInternal("radio.listRadioLocalObjects", postProcessor = Func<JsonElement, obj>(fun d ->
            let mutable ids = JsonElement()
            if d.TryGetProperty("localObjectIds", &ids) then
                ids.EnumerateArray().Select(fun n -> n.GetString()).ToList() :> obj
            else
                List<string>() :> obj))
